deepinv/optim/phase_retrieval.py

```diff
@@ -0,0 +1,112 @@
+import torch
+
+
+def spectral_methods(
+    y: torch.Tensor,
+    physics,
+    x=None,
+    n_iter=50,
+    preprocessing=lambda y: torch.max(1 - 1 / y, torch.tensor(-5.0)),
+    lamb=10.0,
+):
+    r"""
+    Utility function for spectral methods.
+
+    This function runs the Spectral Methods algorithm to find the principal eigenvector of the regularized weighted covariance matrix:
+    
+    .. math::
+        \begin{equation*}
+        M = \conj{B} \text{diag}(T(y)) B + \lambda I,
+        \end{equation*}
+    
+    where :math:`B` is the linear operator of the phase retrieval class, :math:`T(\cdot)` is a preprocessing function for the measurements, and :math:`I` is the identity matrix of corresponding dimensions. Parameter :math:`\lambda` tunes the strength of regularization.
+
+    To find the principal eigenvector, the function runs power iteration which is given by
+
+    .. math::
+        \begin{equation*}
+        \begin{aligned}
+        x_{k+1} &= M x_k \\
+        x_{k+1} &= \frac{x_{k+1}}{\|x_{k+1}\|},
+        \end{aligned}
+        \end{equation*}
+    
+    :param torch.Tensor y: Measurements.
+    :param deepinv.physics physics: Instance of the physics modeling the forward matrix.
+    :param torch.Tensor x: Initial guess for the signals :math:`x_0`.
+    :param int n_iter: Number of iterations.
+    :param function preprocessing: Function to preprocess the measurements. Default is :math:`\max(1 - 1/x, -5)`.
+    :param float lamb: Regularization parameter. Default is 10.
+
+    :return: The estimated signals :math:`x`.
+    """
+    if x is None:
+        x = torch.randn((y.shape[0],) + physics.img_shape, dtype=physics.dtype)
+    x = x.to(torch.cfloat)
+    x = x / torch.linalg.norm(x)
```

Comment:
shouldn't it be a per-element norm (and not over the full batch?)

Reply:
By per-element norm, do you mean x = (x1/||x1||, x2/||x2||, x3/||x3||)? The operation here is to make x have norm 1, not every element of it has norm 1, so I think it is still x = x/||x||.

Reply:
By default deepinv deals with batched data;  i.e. `x` should be a batched tensor of shape `(B, s1, s2, ...)`. Therefore, your method should return a batch of eigenvectors. If B = 1, then what you propose should be fine, but not if B>1. 

Not all functions can be easily batched but I believe that the method you propose can work with batches if you simply replace the normalization?

What I have in mind is the difference with 
```python
torch.manual_seed(0)

x = torch.randn(3, 1, 1, 1)

x_1 = x/torch.linalg.norm(x, keepdim=True)  # Normalizes the full batch
x_2 = x/torch.linalg.norm(x, dim=-1, keepdim=True)  # Normalizes each sub-tensor of the batch

print(x_1.flatten())  # torch.linalg.norm(x_1) = 1
print(x_2.flatten())  # will contain unit elements
```

Reply:
Good idea! I have added batch support for spectral methods.